src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
+
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputingDevice
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
+export AbstractGPUDevice
+
+
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
+select_computing_device(a::ComputingDeviceIndependent, b::AbstractComputingDevice) = b
+select_computing_device(a::AbstractComputingDevice, b::ComputingDeviceIndependent) = a
+
+select_computing_device(a::CPUDevice, ::CPUDevice) = a
+select_computing_device(a::CPUDevice, b::AbstractGPUDevice) = a
+select_computing_device(a::AbstractGPUDevice, b::CPUDevice) = b
```

Comment:
I'd note that this mechanism has a potential extensibility problem (e.g., method ambiguity resolution and associativity/commutativity guarantee). To design this kind of mechanism, I'd look into how it is handled in the broadcast style selection and type promotion rules.

Reply:
This was mainly intended as an internal method to support the "magical" generated default method of `get_computing_device`. If we take that out, I'd suggest to remove `select_computing_device` completely for now.